### PR TITLE
Make `t.timestamps` with precision by default.

### PR DESCRIPTION
**Pull request by kamipo**
I'm not sure it is what we want make it in Rails 6.0.
But I've just implemented to ease to discuss that on this PR.

Context: https://github.com/rails/rails/pull/34956#discussion_r248691413

### CODE REVIEW

File: activerecord/lib/active_record/migration/compatibility.rb [hunk 1]
```diff
@@ -176,13 +218,13 @@ def change_table(table_name, options = {})
...
 
-        def add_reference(*, **options)
+        def add_reference(table_name, ref_name, **options)
           options[:index] ||= false
           super
         end
         alias :add_belongs_to :add_reference
 
-        def add_timestamps(_, **options)
+        def add_timestamps(table_name, **options)
```
Comment (rafaelfranca) [inline activerecord/lib/active_record/migration/compatibility.rb]: Would this not cause warnings of unused arguments?

Comment (kamipo) [reply activerecord/lib/active_record/migration/compatibility.rb]: Currently Ruby doesn't raise any warnings of unused method arguments for now.

https://travis-ci.org/rails/rails/jobs/481187001

**Review by rafaelfranca — APPROVED** [review]

File: actiontext/test/dummy/db/migrate/20180208205311_create_messages.rb [hunk 2]
```diff
@@ -2,7 +2,7 @@ class CreateMessages < ActiveRecord::Migration[6.0]
   def change
     create_table :messages do |t|
       t.string :subject
-      t.timestamps
+      t.timestamps precision: nil
```
Comment (kaspth) [inline actiontext/test/dummy/db/migrate/20180208205311_create_messages.rb]: Why do we need to do this?

File: activerecord/lib/active_record/migration/compatibility.rb [hunk 3]
```diff
@@ -16,13 +16,55 @@ def self.find(version)
...
+            super
+          end
+        end
+
+        def change_table(table_name, **options)
+          if block_given?
+            super { |t| yield compatible_table_definition(t) }
+          else
+            super
+          end
```
Comment (kaspth) [inline activerecord/lib/active_record/migration/compatibility.rb]: Why do we need these similar overrides?

File: activerecord/test/cases/ar_schema_test.rb [hunk 4]
```diff
@@ -142,4 +142,40 @@ def test_timestamps_without_null_set_null_to_false_on_add_timestamps
     assert_not @connection.columns(:has_timestamps).find { |c| c.name == "created_at" }.null
     assert_not @connection.columns(:has_timestamps).find { |c| c.name == "updated_at" }.null
   end
+
+  if subsecond_precision_supported?
```
Comment (kaspth) [inline activerecord/test/cases/ar_schema_test.rb]: Should this be `supports_datetime_with_precision?`?

File: actiontext/test/dummy/db/migrate/20180208205311_create_messages.rb (hunk 2 above)
Comment (kamipo) [reply actiontext/test/dummy/db/migrate/20180208205311_create_messages.rb]: Just keep `db/schema.rb` as before.

File: activerecord/lib/active_record/migration/compatibility.rb (hunk 3 above)
Comment (kamipo) [reply activerecord/lib/active_record/migration/compatibility.rb]: It is to inject `V5_2::TableDefinition` to the table definition instance.

https://github.com/rails/rails/pull/34970/files#diff-2a8be25f82da6b3935cc6a41300a1b01R19`

If the `change_table` is removed, the `test_timestamps_doesnt_set_precision_on_change_table` is failed.

https://github.com/rails/rails/pull/34970/files#diff-ec048630132cce280a9445022318906eR140

**Review by kaspth — APPROVED** [review]

File: actiontext/test/dummy/db/schema.rb [hunk 5]
```diff
@@ -17,8 +17,8 @@
     t.text "body", limit: 16777215
     t.string "record_type", null: false
     t.integer "record_id", null: false
-    t.datetime "created_at", null: false
-    t.datetime "updated_at", null: false
+    t.datetime "created_at", precision: 6, null: false
+    t.datetime "updated_at", precision: 6, null: false
```
Comment (kamipo) [inline actiontext/test/dummy/db/schema.rb]: I've changed to apply new default the same with Action Mailbox.

**Comment by andrewhaines** [comment]
@kamipo Do you think there's any benefit to documenting that this is incompatible w/ MySQL 5.5? I ran across that issue the other day (the shared ClearDB environments for Heroku don't allow higher versions than 5.5) and it took me a while to figure out what was happening. If so, I'm happy to contribute it.

**Comment by kamipo** [comment]
What is the incompatible you mean?

**Comment by andrewhaines** [comment]
`datetime` doesn't support precision in MySQL 5.5.

**Comment by jcoleman** [comment]
@kamipo I'm curious what the reasoning was for this change. Is there something that makes lessening the default precision on databases that support higher precision (e.g., PostgreSQL) for timestamps desirable?

**Comment by jcoleman** [comment]
@kamipo I'm still interested in what the reasoning for this was. Following the links to the discussion it appears this came out of a discussion about MySQL, but this change affected not just MySQL but also databases like Postgres that don't need this change. And I can't find any discussion at all explaining why the change is desirable or was made in the first place.

**Comment by zzak** [comment]
@jcoleman Unfortunately, you might not be able to get a response on a 3 or 4 year old PR. If there is a bug that you've identified, I recommend opening a new issue, or submitting a PR which applies your desired behavior. :pray:

**Comment by jcoleman** [comment]
@zzak Unfortunately getting responses on even new PRs in the Rails project is...not common.

Another [issue](https://github.com/rails/rails/issues/37031) was opened the same year as this PR asking for an explanation of the change, and no one answered the question there either.

**Pull request merged**